File: src/main/java/com/codetopic/utils/container/adapter/ArrayEditAdapter.java
package com.codetopic.utils.container.adapter;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.UiThread;

import com.codetopic.utils.Log;
import com.codetopic.utils.Objects;
import com.codetopic.utils.simple.IteratorWrapper;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;

public abstract class ArrayEditAdapter<T, VH extends UniversalAdapter.ViewHolder>
        extends UniversalAdapter<VH> implements Iterable<T> {

    private static final String LOG_TAG = "ArrayEditAdapter";

    private final Object mLock = new Object();
    private final ArrayList<T> mData = new ArrayList<>();

    public ArrayEditAdapter() {
    }

    public ArrayEditAdapter(Collection<? extends T> data) {
        synchronized (mLock) {
            mData.addAll(data);
        }
    }

    @SafeVarargs
    public ArrayEditAdapter(T... data) {
        synchronized (mLock) {
            Collections.addAll(mData, data);
        }
    }

    @Override
    public int getItemCount() {
        synchronized (mLock) {
            return mData.size();
        }
    }

    @Override
    public T getItem(int position) {
        synchronized (mLock) {
            return mData.get(position);
        }
    }

    public T[] getItems(T[] contents) {
        synchronized (mLock) {
            return mData.toArray(contents);
        }
    }

    public List<T> getItems() {
        synchronized (mLock) {
            //noinspection unchecked
            return (List<T>) mData.clone();
        }
    }

    public int getItemPosition(T item) {
        synchronized (mLock) {
            return mData.indexOf(item);
        }
    }

    @Override
    public boolean isEmpty() {
        synchronized (mLock) {
            return mData.isEmpty();
        }
    }

    @Override
    public Iterator<T> iterator() {
        return new IteratorWrapper<T>(getItems().iterator()) {
            @Override
            public void remove() {
                throw new UnsupportedOperationException("Not supported");
            }
        };
    }

    protected void assertAllowApplyChanges(@Nullable Object editTag,
                                           Collection<Modification<T>> modifications,
                                           @Nullable Collection<T> contentModifiedItems) {
    }

    public Editor<T> edit() {
        return new Editor<>(this);
    }

    @UiThread
    public void postModifications(@NonNull CalculatingMode mode,
                                  Collection<Modification<T>> modifications,
                                  @Nullable Collection<T> contentModifiedItems) {

        postModifications(null, mode, modifications, contentModifiedItems);
    }

    @UiThread
    public void postModifications(@Nullable Object editTag, @NonNull CalculatingMode mode,
                                  Collection<Modification<T>> modifications,
                                  @Nullable Collection<T> contentModifiedItems) {

        assertAllowApplyChanges(editTag, modifications, contentModifiedItems);

        if (!isBaseAttached()) {
            for (Modification<T> modification : modifications)
                modification.modify(null, mData);
            return;
        }
        Base base = getBase();

        synchronized (mLock) {
            try {
                if (mode == CalculatingMode.NO_ANIMATIONS
                        || base.hasOnlySimpleDataChangedReporting()) {
                    for (Modification<T> modification : modifications)
                        modification.modify(null, mData);

                    base.notifyDataSetChanged();
                    return;

                } else if (mode == CalculatingMode.FROM_MODIFICATIONS) {
                    for (Modification<T> modification : modifications)
                        modification.modify(base, mData);

                } else if (mode == CalculatingMode.EQUALS_DETECTION) {
                    //noinspection unchecked
                    List<T> dataBackup = (List<T>) mData.clone();
                    for (Modification<T> modification : modifications)
                        modification.modify(null, mData);

                    boolean oldEmpty = dataBackup.isEmpty();
                    boolean newEmpty = mData.isEmpty();
                    if (!oldEmpty || !newEmpty) {
                        if (oldEmpty) {
                            base.notifyItemRangeInserted(0, mData.size());

                        } else if (newEmpty) {
                            base.notifyItemRangeRemoved(0, dataBackup.size());

                        } else {
                            for (Iterator<T> iterator = dataBackup.iterator(); iterator.hasNext(); ) {
                                T obj = iterator.next();
                                if (!mData.contains(obj)) {
                                    base.notifyItemRemoved(dataBackup.indexOf(obj));
                                    iterator.remove();
                                }
                            }

                            for (int i = 0, size = mData.size(); i < size; i++) {
                                T obj = mData.get(i);
                                if (!dataBackup.contains(obj)) {
                                    dataBackup.add(i, obj);
                                    base.notifyItemInserted(i);
                                    continue;
                                }

                                int oldIndex = dataBackup.indexOf(obj);
                                if (oldIndex != i) {
                                    base.notifyItemMoved(oldIndex, i);
                                    dataBackup.remove(obj);
                                    dataBackup.add(i, obj);
                                }
                            }

                            if (Log.isInDebugMode() && !Objects.equals(dataBackup, mData))
                                Log.e(LOG_TAG, "apply", new InternalError("Detected problem in " + LOG_TAG
                                        + " while applying changes -> !dataBackup.equals(newData)" +
                                        "\ndataBackup: " + dataBackup + "\nmData: " + mData));
                        }
                    }
                } else {
                    throw new IllegalArgumentException("Unknown mode: " + mode);
                }

                if (contentModifiedItems != null) {
                    for (T item : contentModifiedItems) {
                        int index = mData.indexOf(item);
                        if (index != -1) base.notifyItemChanged(index);
                    }
                } else base.notifyItemRangeChanged(0, mData.size());
            } finally {
                onDataEdited(editTag);
            }
        }
    }

    protected void onDataEdited(@Nullable Object editTag) {

    }

    public enum CalculatingMode {
        NO_ANIMATIONS, EQUALS_DETECTION, FROM_MODIFICATIONS
    }

    public interface Modification<T> {
        void modify(@Nullable Base adapterBase, List<T> toModify);
    }

    public static class Editor<T> {

        private static final String LOG_TAG = ArrayEditAdapter.LOG_TAG + "$Editor";

        private final WeakReference<? extends ArrayEditAdapter<T, ?>> mAdapterReference;
        private final ArrayList<Modification<T>> mModifications = new ArrayList<>();
        private final ArrayList<T> mChangedItems = new ArrayList<>();
        private boolean mAllItemsChanged = false;
        private Object mTag = null;

        protected Editor(ArrayEditAdapter<T, ?> adapter) {
            mAdapterReference = new WeakReference<>(adapter);
        }

        @UiThread
        @Nullable
        public <A extends ArrayEditAdapter<T, ?>> A getAdapter() {
            //noinspection unchecked
            return (A) mAdapterReference.get();
        }

        public synchronized Object getTag() {
            return mTag;
        }

        public synchronized Editor<T> setTag(@Nullable Object tag) {
            this.mTag = tag;
            return this;
        }

        public synchronized Editor<T> post(Modification<T> modification) {
            mModifications.add(modification);
            return this;
        }

        public Editor<T> add(final T object) {
            return post(new Modification<T>() {
                @Override
                public void modify(@Nullable Base adapterBase, List<T> toModify) {
                    toModify.add(object);
                    if (adapterBase != null) adapterBase.notifyItemInserted(toModify.size() - 1);
                }
            });
        }

        public Editor<T> add(final int index, final T object) {
            return post(new Modification<T>() {
                @Override
                public void modify(@Nullable Base adapterBase, List<T> toModify) {
                    toModify.add(index, object);
                    if (adapterBase != null) adapterBase.notifyItemInserted(index);
                }
            });
        }

        public Editor<T> addAll(final Collection<? extends T> collection) {
            return post(new Modification<T>() {
                @Override
                public void modify(@Nullable Base adapterBase, List<T> toModify) {
                    toModify.addAll(collection);
                    if (adapterBase != null) {
                        int count = collection.size();
                        adapterBase.notifyItemRangeInserted(toModify.size() - count, count);
                    }
                }
            });
        }

        public Editor<T> addAll(final int index, final Collection<? extends T> collection) {
            return post(new Modification<T>() {
                @Override
                public void modify(@Nullable Base adapterBase, List<T> toModify) {
                    toModify.addAll(index, collection);
                    if (adapterBase != null) adapterBase
                            .notifyItemRangeInserted(index, collection.size());
                }
            });
        }

        public Editor<T> clear() {
            return post(new Modification<T>() {
                @Override
                public void modify(@Nullable Base adapterBase, List<T> toModify) {
                    if (adapterBase == null) {
                        toModify.clear();
                        return;
                    }

                    int count = toModify.size();
                    toModify.clear();
                    adapterBase.notifyItemRangeRemoved(0, count);
                }
            });
        }

        public Editor<T> remove(final int index) {
            return post(new Modification<T>() {
                @Override
                public void modify(@Nullable Base adapterBase, List<T> toModify) {
                    toModify.remove(index);
                    if (adapterBase != null) adapterBase.notifyItemRemoved(index);
                }
            });
        }

        public Editor<T> remove(final Object object) {
            return post(new Modification<T>() {
                @Override
                public void modify(@Nullable Base adapterBase, List<T> toModify) {
                    if (adapterBase == null) {
                        //noinspection SuspiciousMethodCalls
                        toModify.remove(object);
                        return;
                    }

                    //noinspection SuspiciousMethodCalls
                    int index = toModify.indexOf(object);
                    if (index == -1) return;
                    toModify.remove(index);
                    adapterBase.notifyItemRemoved(index);
                }
            });
        }

        public Editor<T> set(final int index, final T object) {
            return post(new Modification<T>() {
                @Override
                public void modify(@Nullable Base adapterBase, List<T> toModify) {
                    toModify.set(index, object);
                    if (adapterBase != null) adapterBase.notifyItemChanged(index);
                }
            });
        }

        public Editor<T> removeAll(final Collection<?> collection) {
            return post(new Modification<T>() {
                @Override
                public void modify(@Nullable Base adapterBase, List<T> toModify) {
                    if (adapterBase == null) {
                        //noinspection SuspiciousMethodCalls
                        toModify.removeAll(collection);
                        return;
                    }

                    List<Integer> ids = new ArrayList<>();
                    for (Object o : collection) {
                        //noinspection SuspiciousMethodCalls
                        int index = toModify.indexOf(o);
                        if (index != -1) ids.add(index);
                    }
                    //noinspection SuspiciousMethodCalls
                    toModify.removeAll(collection);
                    for (Integer i : ids)
                        adapterBase.notifyItemRemoved(i);
                }
            });
        }

        public Editor<T> retainAll(final Collection<?> collection) {
            return post(new Modification<T>() {
                @Override
                public void modify(@Nullable Base adapterBase, List<T> toModify) {
                    if (adapterBase == null) {
                        toModify.retainAll(collection);
                        return;
                    }

                    for (Iterator<T> iterator = toModify.iterator(); iterator.hasNext(); ) {
                        T item = iterator.next();
                        if (!collection.contains(item)) {
                            adapterBase.notifyItemRemoved(toModify.indexOf(item));
                            iterator.remove();
                        }
                    }
                }
            });
        }

        @SafeVarargs
        public synchronized final Editor<T> notifyItemsChanged(T... items) {
            Collections.addAll(mChangedItems, items);
            return this;
        }

        public synchronized Editor<T> notifyItemsChanged(Collection<T> items) {
            mChangedItems.addAll(items);
            return this;
        }

        public synchronized Editor<T> notifyAllItemsChanged() {
            mAllItemsChanged = true;
            return this;
        }

        public synchronized boolean apply() {
            return apply(CalculatingMode.EQUALS_DETECTION);
        }

        @UiThread
        public synchronized boolean apply(@NonNull CalculatingMode mode) {
            ArrayEditAdapter<T, ?> adapter = getAdapter();
            if (adapter != null) {
                adapter.postModifications(getTag(), mode, mModifications,
                        mAllItemsChanged ? null : mChangedItems);
                mModifications.clear();
                mChangedItems.clear();
                mAllItemsChanged = false;
                return true;
            }
            return false;
        }
    }
}

File: src/main/java/eu/codetopic/utils/ui/container/adapter/ArrayEditAdapter.kt
/*
 * utils
 * Copyright (C)   2017  anty
 *
 * This program is free  software: you can redistribute it and/or modify
 * it under the terms  of the GNU General Public License as published by
 * the Free Software  Foundation, either version 3 of the License, or
 * (at your option) any  later version.
 *
 * This program is distributed in the hope that it  will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied  warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
 * GNU General Public License for more details.
 *
 * You  should have received a copy of the GNU General Public License
 * along  with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package eu.codetopic.utils.ui.container.adapter

import android.support.annotation.UiThread

import java.lang.ref.WeakReference
import java.util.Collections

import eu.codetopic.java.utils.log.Log
import eu.codetopic.utils.ui.container.adapter.ArrayEditAdapter.CalculatingMode.*
import eu.codetopic.java.utils.forEachIterate
import eu.codetopic.java.utils.debug.DebugMode

abstract class ArrayEditAdapter<E : Any, VH : UniversalHolder<*>>() :
        UniversalAdapter<VH>(), Iterable<E> {

    companion object {

        private const val LOG_TAG = "ArrayEditAdapter"
    }

    private val dataLock = Any()
    private val data = mutableListOf<E>()

    override val itemCount: Int
        get() = synchronized(dataLock) { data.size }

    open val items: List<E>
        get() = synchronized(dataLock) { data.toList() }

    override val isEmpty: Boolean
        get() = synchronized(dataLock) { data.isEmpty() }

    constructor(data: Collection<E>) : this() {
        this.data.addAll(data)
    }

    constructor(vararg data: E) : this() {
        Collections.addAll(this.data, *data)
    }

    override fun getItem(position: Int): E = synchronized(dataLock) { data[position] }

    open fun getItemPosition(item: E): Int = synchronized(dataLock) { data.indexOf(item) }

    override fun iterator(): Iterator<E> = object : Iterator<E> by items.iterator() {}

    fun edit(): Editor<E> = Editor(this)

    protected open fun assertAllowApplyChanges(editTag: Any?,
                                               modifications: Collection<Modification<E>>,
                                               contentModifiedItems: Collection<E>?) {
    }

    @UiThread
    open fun postModifications(mode: CalculatingMode,
                               modifications: Collection<Modification<E>>,
                               contentModifiedItems: Collection<E>?) {

        postModifications(null, mode, modifications, contentModifiedItems)
    }

    @UiThread
    open fun postModifications(editTag: Any?, mode: CalculatingMode,
                               modifications: Collection<Modification<E>>,
                               contentModifiedItems: Collection<E>?) {

        assertAllowApplyChanges(editTag, modifications, contentModifiedItems)

        synchronized(dataLock) {
            if (!isBaseAttached) return modifications.forEach { it(data, null) }

            val realMode = mode.takeIf { !base.hasOnlySimpleDataChangedReporting() } ?: NO_ANIMATIONS

            try {
                when (realMode) {
                    NO_ANIMATIONS -> {
                        modifications.forEach { it(data, null) }
                        base.notifyDataSetChanged()
                        return
                    }
                    FROM_MODIFICATIONS -> {
                        val base = base
                        modifications.forEach { it(data, base) }
                    }
                    EQUALS_DETECTION -> {
                        val dataBackup = data.toMutableList()

                        modifications.forEach { it(data, null) }

                        val oldEmpty = dataBackup.isEmpty()
                        val newEmpty = data.isEmpty()
                        when {
                            oldEmpty && newEmpty -> {
                            }
                            oldEmpty -> base.notifyItemRangeInserted(0, data.size)
                            newEmpty -> base.notifyItemRangeRemoved(0, dataBackup.size)
                            else -> {
                                // Cache base, don't get it every time when some change is found
                                val base = base

                                dataBackup.forEachIterate { iterator, obj ->
                                    if (!data.contains(obj)) {
                                        base.notifyItemRemoved(dataBackup.indexOf(obj))
                                        iterator.remove()
                                    }
                                }

                                data.forEachIndexed forI@ { i, obj ->
                                    if (!dataBackup.contains(obj)) {
                                        dataBackup.add(i, obj)
                                        base.notifyItemInserted(i)
                                        return@forI
                                    }

                                    val oldIndex = dataBackup.indexOf(obj)
                                    if (oldIndex != i) {
                                        base.notifyItemMoved(oldIndex, i)
                                        dataBackup.remove(obj)
                                        dataBackup.add(i, obj)
                                    }
                                }

                                if (DebugMode.isEnabled && dataBackup != data)
                                    Log.e(LOG_TAG, "apply", InternalError("Detected problem " +
                                            "in $LOG_TAG while applying changes (dataBackup != newData) ->" +
                                            "\n(dataBackup=$dataBackup,\nnewData=$data"))
                            }
                        }
                    }
                    // else -> throw IllegalArgumentException("Unknown mode -> (mode=$mode)")
                }

                if (realMode != NO_ANIMATIONS) {
                    if (contentModifiedItems != null) {
                        val base = base // Cache base, don't get it every iteration
                        contentModifiedItems.mapNotNull {
                            data.indexOf(it).takeIf { it != -1 }
                        }.forEach { base.notifyItemChanged(it) }
                    } else base.notifyItemRangeChanged(0, data.size)
                }
            } finally {
                onDataEdited(editTag)
            }
        }
    }

    protected open fun onDataEdited(editTag: Any?) {

    }

    enum class CalculatingMode {
        NO_ANIMATIONS, EQUALS_DETECTION, FROM_MODIFICATIONS
    }

    class Editor<E : Any>(adapter: ArrayEditAdapter<E, *>) {

        companion object {

            private const val LOG_TAG = "${ArrayEditAdapter.LOG_TAG}\$Editor"
        }

        private val adapterReference: WeakReference<out ArrayEditAdapter<E, *>> = WeakReference(adapter)
        private val modifications = mutableListOf<Modification<E>>()
        private val changedItems = mutableListOf<E>()
        private var allItemsChanged = false

        @get:Synchronized
        @set:Synchronized
        var tag: Any? = null

        @Synchronized
        fun post(modification: Modification<E>) {
            modifications.add(modification)
        }

        fun add(obj: E) = post {
            add(obj)
            it?.notifyItemInserted(size - 1)
        }

        fun add(index: Int, obj: E) = post {
            add(index, obj)
            it?.notifyItemInserted(index)
        }

        fun <AT : E> addAll(array: Array<AT>) = addAll(array.asList())

        fun addAll(collection: Collection<E>) = post {
            addAll(collection)
            it?.apply {
                val count = collection.size
                notifyItemRangeInserted(size - count, count)
            }
        }

        fun addAll(index: Int, collection: Collection<E>) = post {
            addAll(index, collection)
            it?.notifyItemRangeInserted(index, collection.size)
        }

        fun clear() = post {
            if (it == null) {
                clear()
                return@post
            }

            val count = size
            clear()
            it.notifyItemRangeRemoved(0, count)
        }

        fun removeAt(index: Int) = post {
            removeAt(index)
            it?.notifyItemRemoved(index)
        }

        fun remove(obj: Any) = post {
            if (it == null) {
                remove(obj)
                return@post
            }

            indexOf(obj).takeIf { it != -1 }?.apply {
                removeAt(this)
                it.notifyItemRemoved(this)
            }
        }

        operator fun set(index: Int, obj: E) = post {
            set(index, obj)
            it?.notifyItemChanged(index)
        }

        fun removeAll(collection: Collection<*>) = post {
            if (it == null) {
                removeAll(collection)
                return@post
            }

            collection.mapNotNull {
                indexOf(it).takeIf { it != -1 }
            }.forEach { index ->
                removeAt(index)
                it.notifyItemRemoved(index)
            }
        }

        fun retainAll(collection: Collection<*>) = post {
            if (it == null) {
                retainAll(collection)
                return@post
            }

            forEachIterate { iterator, obj ->
                if (!collection.contains(obj)) {
                    it.notifyItemRemoved(indexOf(obj))
                    iterator.remove()
                }
            }
        }

        @Synchronized
        fun notifyItemsChanged(vararg items: E) {
            Collections.addAll(changedItems, *items)
        }

        @Synchronized
        fun notifyItemsChanged(items: Collection<E>) {
            changedItems.addAll(items)
        }

        @Synchronized
        fun notifyAllItemsChanged(){
            allItemsChanged = true
        }

        @Synchronized
        fun apply(): Boolean = apply(CalculatingMode.EQUALS_DETECTION)

        @UiThread
        @Synchronized
        fun apply(mode: CalculatingMode): Boolean {
            return adapterReference.get()?.let {
                it.postModifications(tag, mode, modifications,
                        if (allItemsChanged) null else changedItems)
                modifications.clear()
                changedItems.clear()
                allItemsChanged = false

                true
            } ?: false
        }
    }
}
